### PR TITLE
add main logic for script (not tested)

### PR DESCRIPTION
**Pull request by maxlou05**

### CODE REVIEW

**Review by mgupta27 — APPROVED** [review]
Conditionally approve.

Please write a rc.local.copyme that contains the command to run the code.

File: rc.local.copyme [hunk 1]
```diff
@@ -0,0 +1 @@
+python3 signal_strength_tracker.py &
```
Comment (Xierumeng) [inline rc.local.copyme]: Is the user home directory the default for `rc.local` ? I recommend an absolute path.

Also, virtual environment?

File: requirements.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# Packages listed in alphabetical order
+
```
Comment (Xierumeng) [inline requirements.txt]: Remove empty line.

File: signal_strength_tracker.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+
+# Parameters
+SERIAL_PORT = "/dev/ttyUSB3"  # Change this to match your modem's serial port
+BAUDRATE = 115200  # Change this to match your modem's baud rate
+AT_COMMAND = b"AT+CSQ\r\n"  # Command to run on modem
+OUTPUT_LOG_FILE_PATH = pathlib.Path(f"/media/warg/E3F9-7EE3/signal_log_{int(time.time())}.txt")
+PERIOD = 3
+
+# Wait 20 seconds becasue the RPi has a setting where it doesn't boot until 20 seconds after power
```
Comment (Xierumeng) [inline signal_strength_tracker.py]: Add a main guard.

File: signal_strength_tracker.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+time.sleep(20)
```
Comment (Xierumeng) [inline signal_strength_tracker.py]: Make this value a global constant.

**Review by Xierumeng — COMMENTED** [review]
Reviewed.

File: signal_strength_tracker.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+Logs LTE signal strength into a file.
+"""
+import datetime
+import pathlib
+import time
+
+import serial
+
+
+# Parameters
```
Comment (Xierumeng) [inline signal_strength_tracker.py]: Remove, unnecessary.

File: signal_strength_tracker.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+SERIAL_PORT = "/dev/ttyUSB3"  # Change this to match your modem's serial port
```
Comment (Xierumeng) [inline signal_strength_tracker.py]: Remove 2nd person references (i.e. replace with `the` ).

Comment (Xierumeng) [inline signal_strength_tracker.py]: Is modem referring to the LTE device?

File: signal_strength_tracker.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+BAUDRATE = 115200  # Change this to match your modem's baud rate
```
Comment (Xierumeng) [inline signal_strength_tracker.py]: Same as above.

File: signal_strength_tracker.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+AT_COMMAND = b"AT+CSQ\r\n"  # Command to run on modem
```
Comment (Xierumeng) [inline signal_strength_tracker.py]: What does this command do?

File: signal_strength_tracker.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+OUTPUT_LOG_FILE_PATH = pathlib.Path(f"/media/warg/E3F9-7EE3/signal_log_{int(time.time())}.txt")
```
Comment (Xierumeng) [inline signal_strength_tracker.py]: Separate each directory as an argument. For example:
```py
pathlib.Path("/", "absolute", "path", "to", "file.txt")
```

File: signal_strength_tracker.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+PERIOD = 3
```
Comment (Xierumeng) [inline signal_strength_tracker.py]: Add empty line above this line.

Comment (Xierumeng) [inline signal_strength_tracker.py]: Add comment: `# seconds`

File: signal_strength_tracker.py [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+SETUP_TIME = 20
```
Comment (Xierumeng) [inline signal_strength_tracker.py]: Add comment: `# seconds`

Comment (Xierumeng) [inline signal_strength_tracker.py]: Move this above line 16.

File: signal_strength_tracker.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
```
Comment (Xierumeng) [inline signal_strength_tracker.py:27]: 2 empty lines between global constants and top level functions.

File: signal_strength_tracker.py [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+if __name__ == "__main__":
+    # Wait 20 seconds becasue the RPi has a setting where it doesn't boot until 20 seconds after power
```
Comment (Xierumeng) [inline signal_strength_tracker.py]: Remove or move to the constant.

Comment (Xierumeng) [inline signal_strength_tracker.py]: Is it the Raspberry Pi itself that doesn't boot for 20 seconds? If the Raspberry Pi isn't booted how is the program supposed to run? Is it something else that isn't booting immediately?

File: signal_strength_tracker.py [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    time.sleep(SETUP_TIME)
+
+    while True:
+        log_file = open(OUTPUT_LOG_FILE_PATH, 'a')
```
Comment (Xierumeng) [inline signal_strength_tracker.py]: Add: `, encoding="utf-8"`

File: signal_strength_tracker.py [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+
+        # Create serial connection
+        ser = serial.Serial(
+            SERIAL_PORT,
+            baudrate=BAUDRATE,
+            timeout=1,
```
Comment (Xierumeng) [inline signal_strength_tracker.py]: Make this a global constant.

File: signal_strength_tracker.py [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+        )
+        
+        # Write AT command to the modem
+        ser.write(AT_COMMAND)
+
+        # Get result back from the modem (reads 1000 bytes)
+        response = ser.read(1000)
```
Comment (Xierumeng) [inline signal_strength_tracker.py]: Make this a global constant.

File: signal_strength_tracker.py [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+
+        # String parsing can be done here if you want to edit the log file output
+        log_file.write(
+            datetime.datetime.now().isoformat(timespec="seconds") + \
+            "    Response: " + response.decode('utf-8') + "\n"
```
Comment (Xierumeng) [inline signal_strength_tracker.py]: Use double quotation marks.

File: signal_strength_tracker.py [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+        )
+
+        log_file.close()
+        ser.close()
```
Comment (Xierumeng) [inline signal_strength_tracker.py:54]: Any particular reason why the file and serial port is being opened and reclosed on every iteration of the loop?

**Review by Xierumeng — COMMENTED** [review]
Reviewed.

File: rc.local.copyme (hunk 1 above)
Comment (Xierumeng) [inline rc.local.copyme]: See the other repositories for better examples of `rc.local.copyme` .

**Review by Xierumeng — CHANGES_REQUESTED** [review]
I am blocking the merge until the environment is resolved.

File: rc.local.copyme [hunk 19]
```diff
@@ -0,0 +1,3 @@
+# Copy the following contents to the end of rc.local:
+
+python3 /home/warg/signal_strength_tracker.py &
```
Comment (Xierumeng) [inline rc.local.copyme]: Add virtual environment activation.

File: signal_strength_tracker.py [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+BAUDRATE = 115200  # Change this to match the LTE modem's baud rate
+AT_COMMAND = b"AT+CSQ\r\n"  # Command to run on modem (returns signal strength)
+OUTPUT_LOG_FILE_PATH = pathlib.Path(
+    "/",
+    "media",
+    "warg",
+    "E3F9-7EE3",
+    f"signal_log_{int(time.time())}.txt"
+)
+TIMEOUT = 1  # Seconds
```
Comment (Xierumeng) [inline signal_strength_tracker.py]: More descriptive variable name (e.g. `SERIAL_INITIALIZATION_TIMEOUT` ).

File: signal_strength_tracker.py [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+BYTES_TO_READ = 1000  # Maximum bytes to read after command is run
```
Comment (Xierumeng) [inline signal_strength_tracker.py]: More descriptive variable name (e.g. `MAX_BYTES_TO_READ` ).

File: signal_strength_tracker.py [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+
+SETUP_TIME = 20  # Wait 20s since the RPi has a setting where it doesn't boot until 20s after power
```
Comment (Xierumeng) [inline signal_strength_tracker.py]: Where is this setting? Add the information to the Confluence documentation and reference it (e.g. `See documentation for more detail` ).

**Review by Xierumeng — COMMENTED** [review]
Reviewed.

File: rc.local.copyme [hunk 23]
```diff
@@ -0,0 +1,5 @@
+# Copy the following contents to the end of rc.local:
+
+cd /home/warg/venv
```
Comment (Xierumeng) [inline rc.local.copyme]: `cd /home/warg/signal-strength-logging/`

File: rc.local.copyme [hunk 24]
```diff
@@ -0,0 +1,5 @@
...
+source bin/activate
```
Comment (Xierumeng) [inline rc.local.copyme]: `source venv/bin/activate`

File: rc.local.copyme [hunk 25]
```diff
@@ -0,0 +1,5 @@
...
+python3 /home/warg/signal_strength_tracker.py &
```
Comment (Xierumeng) [inline rc.local.copyme]: `python -m signal-strength-tracker`

`&` optional.

**Review by Xierumeng — COMMENTED** [review]
Reviewed. Alternatively, create `run.sh` to be called.

**Review by Xierumeng — APPROVED** [review]
Approved.

**Pull request merged**